### PR TITLE
openvpn auth poc with some cli fixes

### PR DESCRIPTION
**Pull request by zolia**

### CODE REVIEW

File: cmd/mysterium_client/cli/command.go [hunk 1]
```diff
@@ -147,7 +147,7 @@ func (c *Command) connect(line string) {
 }
 
 func (c *Command) unlock(line string) {
-	unlockArgs := strings.TrimSpace(line[7:])
+	unlockArgs := strings.TrimSpace(line[6:])
```
Comment (donce) [inline cmd/mysterium_client/cli/command.go]: There is quite a lot of boiler-plate code in cli - this error might have been avoided by having "unlock" string const and using i.e. `strings.TrimSpace(line[len(UNLOCK)])`. But that's beyond scope of this PR.

File: openvpn/middlewares/server/auth/middleware.go [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+	"strconv"
+)
+
+type middleware struct {
+	authenticator Authenticator
+	connection    net.Conn
+	username      string
+	password      string
+	clientId      int
+	keyId         int
```
Comment (donce) [inline openvpn/middlewares/server/auth/middleware.go]: it should be `clientID`, `keyID`

File: openvpn/middlewares/client/auth/middleware.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+// NewMiddleware creates clint user_auth challenge authentication middleware
+func NewMiddleware(authenticator Authenticator) openvpn.ManagementMiddleware {
+	return &middleware{
+		authenticator: authenticator,
+		connection:    nil,
+	}
+}
+
+func (m *middleware) Start(connection net.Conn) error {
+	m.connection = connection
```
Comment (donce) [inline openvpn/middlewares/client/auth/middleware.go]: `connection` can be passed in in `NewMiddleware` - we'll have less mutability then.

File: openvpn/middlewares/server/auth/middleware.go [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+	state         middlewares.State
```
Comment (donce) [inline openvpn/middlewares/server/auth/middleware.go]: Are all fields required to be struct fields? I.e. `password` seems only needed in `ConsumeLine` method - limiting it's scope to that method might simplify this struct.

Comment (zolia) [reply openvpn/middlewares/server/auth/middleware.go]: ConsumeLine works by analysing line by line. Thus it needs to store some parsed data as a state to be used for later time by the same method.

File: openvpn/middlewares/client/auth/middleware.go (hunk 3 above)
Comment (zolia) [reply openvpn/middlewares/client/auth/middleware.go]: according to current structure, Start() is called from management.Start() method, where listener is created. ManagementMiddleware interface defines this. We cannot change without reworking all implementations.

File: openvpn/middlewares/server/auth/middleware.go [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+func (m *middleware) Stop() error {
+	_, err := m.connection.Write([]byte("state off all\n"))
+	return err
+}
+
+func (m *middleware) State() middlewares.State {
+	return m.state
+}
+
+func (m *middleware) ConsumeLine(line string) (consumed bool, err error) {
```
[the remark this answers is not in the document]
Comment (zolia) [reply openvpn/middlewares/server/auth/middleware.go]: true true, did split off some inner function from it.

File: openvpn/middlewares/client/auth/middleware.go (hunk 3 above)
Comment (donce) [reply openvpn/middlewares/client/auth/middleware.go]: True, lets keep it as it is for now :)

File: openvpn/middlewares/client/auth/authenticator.go [hunk 6]
```diff
@@ -0,0 +1,14 @@
+package auth
+
+// Authenticator returns client's current auth primitives (i.e. customer identity signature / node's sessionId)
+type Authenticator func() (username string, password string, err error)
+
+// NewAuthenticator returns Authenticator callback
+func NewAuthenticator() Authenticator {
```
Comment (Waldz) [inline openvpn/middlewares/client/auth/authenticator.go]: `NewAuthenticatorFake`

File: openvpn/middlewares/client/auth/middleware_test.go [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+	middleware := NewMiddleware(authenticator.auth)
+	assert.NotNil(t, middleware)
+}
+
+func Test_ConsumeLineSkips(t *testing.T) {
+	var tests = []struct {
+		line string
+	}{
+		{">SOME_LINE_DELIVERED"},
+		{">ANOTHER_LINE_DELIVERED"},
```
Comment (Waldz) [inline openvpn/middlewares/client/auth/middleware_test.go]: `PASSWORD:`

File: openvpn/middlewares/client/auth/middleware_test.go [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+	authenticator := &fakeAuthenticator{}
+	middleware := NewMiddleware(authenticator.auth)
+	connection := &fakeConnection{}
+	middleware.Start(connection)
+
+	for _, test := range tests {
+		consumed, err := middleware.ConsumeLine(test.line)
+		assert.NoError(t, err, test.line)
+		assert.True(t, consumed, test.line)
```
Comment (Waldz) [inline openvpn/middlewares/client/auth/middleware_test.go]: Should be asserting values, passed to `authenticator`

File: openvpn/middlewares/server/auth/authenticator.go [hunk 9]
```diff
@@ -0,0 +1,16 @@
+package auth
+
+// Authenticator callback checks given auth primitives (i.e. customer identity signature / node's sessionId)
+type Authenticator func(username, password string) (bool, error)
+
+// NewAuthenticator returns Authenticator callback
+func NewAuthenticator() Authenticator {
```
Comment (Waldz) [inline openvpn/middlewares/server/auth/authenticator.go]: 2 names are the same. What about `AuthChecker` or `AuthService`?

File: openvpn/middlewares/server/auth/middleware.go [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+	_, err := m.connection.Write([]byte("state on all\n"))
+	return err
+}
+
+func (m *middleware) Stop() error {
+	_, err := m.connection.Write([]byte("state off all\n"))
+	return err
+}
+
+func (m *middleware) State() middlewares.State {
```
Comment (Waldz) [inline openvpn/middlewares/server/auth/middleware.go]: How can server have one single state? It has state for each client, is not it?

File: openvpn/middlewares/server/auth/middleware.go [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+		}
+		return true, nil
+	}
+
+	return false, nil
+}
+func (m *middleware) Reset() {
+	m.username = ""
+	m.password = ""
+	m.clientId = -1
```
Comment (Waldz) [inline openvpn/middlewares/server/auth/middleware.go]: Strange value, I like -69 more

File: openvpn/middlewares/server/auth/middleware.go [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+		return true, nil
+	}
+
+	rule, err = regexp.Compile("^>CLIENT:ENV,END$")
+	if err != nil {
+		return false, err
+	}
+
+	match = rule.FindStringSubmatch(line)
+	if len(match) > 0 {
```
Comment (Waldz) [inline openvpn/middlewares/server/auth/middleware.go]: Function too long, at least each state action could be moved to different functions

File: openvpn/middlewares/server/auth/middleware_test.go [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+	var tests = []struct {
+		line            string
+		unexpectedState middlewares.State
+	}{
+		{">CLIENT:ENV,password=12341234", middlewares.STATE_AUTH},
+		{">CLIENT:ENV,username=username", middlewares.STATE_AUTH},
+	}
+
+	for _, test := range tests {
+		authenticator := NewFakeAuthenticator()
```
Comment (Waldz) [inline openvpn/middlewares/server/auth/middleware_test.go]: I miss state there `authentificationChecker` has mocked login result

File: openvpn/middlewares/client/auth/middleware.go [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+	"github.com/mysterium/node/openvpn"
+	"github.com/mysterium/node/openvpn/middlewares"
+	"net"
+	"regexp"
+)
+
+type middleware struct {
+	authenticator Authenticator
+	connection    net.Conn
+	username      string
```
[the remark this answers is not in the document]
Comment (zolia) [reply openvpn/middlewares/client/auth/middleware.go]: yes, that determines only current session being authorised. You cannot perform more than 1 authorisation  at the same time via openvpn management interface. Access is serial, thus we only care about last (current) user. Will rename that.

File: openvpn/middlewares/server/auth/middleware.go (hunk 10 above)
Comment (zolia) [reply openvpn/middlewares/server/auth/middleware.go]: No, management interface processes all incoming request serially. Looks like auth calls are atomic.

File: openvpn/middlewares/client/bytescount/middleware.go [hunk 15]
```diff
@@ -39,6 +41,10 @@ func (middleware *middleware) Stop() error {
 	return err
 }
 
+func (middleware *middleware) State() middlewares.State {
```
Comment (Waldz) [inline openvpn/middlewares/client/bytescount/middleware.go]: We should not expose private variable publicly

File: openvpn/middlewares/client/auth/middleware.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (Waldz) [reply openvpn/middlewares/client/auth/middleware.go]: If so, we cant expose these internals to public methods. I mean `middleware.State()` function

File: openvpn/middlewares/server/auth/middleware_test.go (hunk 13 above)
Comment (zolia) [reply openvpn/middlewares/server/auth/middleware_test.go]: Should we test AuthenticatorChecker in authenticator_test.go separately ? middleware exposes only ConsumeLine method which does not give you authentication outcome result.

File: openvpn/middlewares/server/auth/middleware_test.go [hunk 16]
```diff
@@ -0,0 +1,228 @@
...
+	var tests = []struct {
+		line string
+	}{
+		{">CLIENT:REAUTH,0,0"},
+		{">CLIENT:CONNECT,0,0"},
+		{">CLIENT:ENV,password=12341234"},
+		{">CLIENT:ENV,username=username"},
+	}
+
+	var fac FakeAuthenticatorChecker
```
Comment (Waldz) [inline openvpn/middlewares/server/auth/middleware_test.go]: Why this variable is needed? Usualy You just call call factory function `package.New...()` in package

File: openvpn/middlewares/client/auth/middleware_test.go [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+}
+
+func (conn *fakeConnection) Write(b []byte) (n int, err error) {
+	conn.lastDataWritten = b
+	return 0, nil
+}
+
+func Test_Factory(t *testing.T) {
+	authenticator := &fakeAuthenticator{}
+	middleware := NewMiddleware(authenticator.auth)
```
Comment (Waldz) [inline openvpn/middlewares/client/auth/middleware_test.go]: You dont need struct `fakeAuthenticator` here, just pass `auth` callback

File: openvpn/middlewares/server/auth/middleware_test.go [hunk 18]
```diff
@@ -0,0 +1,228 @@
...
+}
+
+func Test_ConsumeLineSkips(t *testing.T) {
+	var tests = []struct {
+		line string
+	}{
+		{">SOME_LINE_TO_BE_DELIVERED"},
+		{">ANOTHER_LINE_TO_BE_DELIVERED"},
+	}
+	var fac FakeAuthenticatorChecker
```
Comment (Waldz) [inline openvpn/middlewares/server/auth/middleware_test.go]: Unneeded struct here also

File: openvpn/management.go [hunk 19]
```diff
@@ -26,14 +26,15 @@ type ManagementMiddleware interface {
 	Start(connection net.Conn) error
 	Stop() error
 	ConsumeLine(line string) (consumed bool, err error)
+	State() (state State)
```
Comment (Waldz) [inline openvpn/management.go]: Interface is collapsing

File: openvpn/middlewares/server/auth/middleware_test.go [hunk 20]
```diff
@@ -0,0 +1,228 @@
...
+		var fac FakeAuthenticatorChecker
+		authenticator := fac.NewFakeAuthenticatorChecker()
+		middleware := NewMiddleware(authenticator.authenticator)
+		connection := &fakeConnection{}
+		middleware.Start(connection)
+
+		consumed, err := middleware.ConsumeLine(test.line)
+		assert.NoError(t, err, test.line)
+		assert.True(t, consumed, test.line)
+		assert.Equal(t, test.expectedState, middleware.State())
```
Comment (Waldz) [inline openvpn/middlewares/server/auth/middleware_test.go]: Lets not expose internal `state`, your test can read it from private attrribute `middleware.state`

File: openvpn/middlewares/server/auth/middleware_test.go [hunk 21]
```diff
@@ -0,0 +1,228 @@
...
+		expectedState openvpn.State
+	}{
+		{">CLIENT:REAUTH,0,0", openvpn.STATE_AUTH},
+		{">CLIENT:CONNECT,0,0", openvpn.STATE_AUTH},
+	}
+
+	for _, test := range tests {
+		var fac FakeAuthenticatorChecker
+		authenticator := fac.NewFakeAuthenticatorChecker()
+		middleware := NewMiddleware(authenticator.authenticator)
```
Comment (Waldz) [inline openvpn/middlewares/server/auth/middleware_test.go]: In tests I suggest construct middleware without factory `NewMiddleware()` - gives more control

File: openvpn/middlewares/client/auth/middleware.go [hunk 22]
```diff
@@ -0,0 +1,73 @@
...
+type middleware struct {
+	authenticator Authenticator
+	connection    net.Conn
+	lastUsername  string
+	lastPassword  string
+	state         openvpn.State
+}
+
+// NewMiddleware creates clint user_auth challenge authentication middleware
+func NewMiddleware(authenticator Authenticator) openvpn.ManagementMiddleware {
```
Comment (Waldz) [inline openvpn/middlewares/client/auth/middleware.go]: Return more exact type of your return `*middleware`

File: openvpn/middlewares/server/auth/middleware.go [hunk 23]
```diff
@@ -0,0 +1,203 @@
...
+	connection    net.Conn
+	lastUsername  string
+	lastPassword  string
+	clientID      int
+	keyID         int
+	state         openvpn.State
+}
+
+// NewMiddleware creates server user_auth challenge authentication middleware
+func NewMiddleware(authenticator AuthenticatorChecker) openvpn.ManagementMiddleware {
```
Comment (Waldz) [inline openvpn/middlewares/server/auth/middleware.go]: Return more exact type of your return `*middleware`

File: openvpn/management.go (hunk 19 above)
Comment (Waldz) [reply openvpn/management.go]: You need State() only in tests. So construct exact structure (not interface) there

File: bin/helpers/functions.sh [hunk 24]
```diff
@@ -4,6 +4,6 @@
 function get_linker_ldflags {
     echo "
         -X 'github.com/mysterium/node/server.mysteriumApiUrl=${MYSTERIUM_API_URL}'
```
Comment (Waldz) [inline bin/helpers/functions.sh]: Not the same with `mysteriumApiUrl`?

Comment (donce) [reply bin/helpers/functions.sh]: No, that was was not renamed in code yet.

File: openvpn/config.go [hunk 25]
```diff
@@ -47,8 +47,11 @@ func (c *Config) SetDevice(deviceName string) {
 	c.setParam("dev", deviceName)
 }
 
-func (c *Config) SetTlsCertificate(caFile, certFile, certKeyFile string) {
+func (c *Config) SetTlsCACertificate(caFile string) {
```
Comment (donce) [inline openvpn/config.go]: from linter:
>method SetTlsCACertificate should be SetTLSCACertificate

File: openvpn/middlewares/client/state/middleware_test.go [hunk 26]
```diff
@@ -1,13 +1,14 @@
...
+	"github.com/mysterium/node/openvpn"
 	"github.com/stretchr/testify/assert"
 	"testing"
 )
 
-var currentState State
+var currentState openvpn.State
 
-func updateCurrentState(state State) error {
+func updateCurrentState(state openvpn.State) error {
```
[the remark this answers is not in the document]
Comment (zolia) [reply openvpn/middlewares/client/state/middleware_test.go]: not my code, but I see its used later in state middleware Test_ConsumeLineTakes

File: openvpn/config.go [hunk 27]
```diff
@@ -47,8 +47,11 @@ func (c *Config) SetDevice(deviceName string) {
...
 	c.AddOptions(OptionFile("ca", caFile))
+}
+
+func (c *Config) SetTlsPrivatePubKeys(certFile string, certKeyFile string) {
```
Comment (donce) [inline openvpn/config.go]: from linter:

>  method SetTlsPrivatePubKeys should be SetTLSPrivatePubKeys

File: openvpn/middlewares/client/auth/middleware.go [hunk 28]
```diff
@@ -0,0 +1,73 @@
...
+
+type middleware struct {
+	authenticator Authenticator
+	connection    net.Conn
+	lastUsername  string
+	lastPassword  string
+	state         openvpn.State
+}
+
+// NewMiddleware creates clint user_auth challenge authentication middleware
```
Comment (donce) [inline openvpn/middlewares/client/auth/middleware.go]: cli**e**nt

File: cmd/mysterium_server/command_run/factory.go [hunk 29]
```diff
@@ -81,7 +80,11 @@ func NewCommandWith(
 				filepath.Join(options.DirectoryConfig, "crl.pem"),
 				filepath.Join(options.DirectoryConfig, "ta.key"),
 			)
-			return openvpn.NewServer(vpnServerConfig, options.DirectoryRuntime)
+			authenticator := auth.NewAuthenticatorFake()
+			vpnMiddlewares := []openvpn.ManagementMiddleware{
+				auth.NewMiddleware(authenticator),
+			}
+			return openvpn.NewServer(vpnServerConfig, options.DirectoryRuntime, vpnMiddlewares...)
```
[the remark this answers is not in the document]
Comment (zolia) [reply cmd/mysterium_server/command_run/factory.go]: its more general case, just like in client factory. There might be more than 1 middleware initialised.

[the remark this answers is not in the document]
Comment (Waldz) [reply cmd/mysterium_server/command_run/factory.go]: Array is needed then appending middlewares dynamically

File: openvpn/middlewares/client/bytescount/middleware.go [hunk 30]
```diff
@@ -39,6 +40,10 @@ func (middleware *middleware) Stop() error {
 	return err
 }
 
+func (middleware *middleware) State() openvpn.State {
```
Comment (Waldz) [inline openvpn/middlewares/client/bytescount/middleware.go]: This middleware doest not have state at all

File: openvpn/middlewares/client/state/middleware.go [hunk 31]
```diff
@@ -9,9 +9,10 @@ import (
 type middleware struct {
 	listeners  []clientStateCallback
 	connection net.Conn
+	state      openvpn.State
```
Comment (Waldz) [inline openvpn/middlewares/client/state/middleware.go]: Unused `state` attribute

File: openvpn/middlewares/server/auth/authenticator.go [hunk 32]
```diff
@@ -0,0 +1,16 @@
+package auth
+
+// AuthenticatorChecker callback checks given auth primitives (i.e. customer identity signature / node's sessionId)
+type AuthenticatorChecker func(username, password string) (bool, error)
+
+// NewAuthenticatorFake returns AuthenticatorChecker callback
+func NewAuthenticatorFake() AuthenticatorChecker {
```
Comment (Waldz) [inline openvpn/middlewares/server/auth/authenticator.go]: Dont repeat package name in subnames:
- `authenticator.go` -> `checker.go`
- `NewAuthenticatorFake()` -> NewCheckerFake()

File: openvpn/middlewares/server/auth/middleware_test.go [hunk 33]
```diff
@@ -0,0 +1,191 @@
...
+	connection := &fakeConnection{}
+	middleware.Start(connection)
+
+	for _, test := range tests {
+		consumed, err := middleware.ConsumeLine(test.line)
+		assert.NoError(t, err, test.line)
+		assert.True(t, consumed, test.line)
+		assert.Equal(t, test.expectedState, middleware.state)
+	}
+	assert.Equal(t, "username1", middleware.lastUsername)
```
Comment (Waldz) [inline openvpn/middlewares/server/auth/middleware_test.go]: Should assert if `authenticator` was called + has  this username&password pair

File: openvpn/middlewares/server/auth/middleware_test.go [hunk 34]
```diff
@@ -0,0 +1,191 @@
...
+
+	for _, test := range tests {
+		middleware := NewMiddleware(fakeAuthenticator)
+		connection := &fakeConnection{}
+		middleware.Start(connection)
+
+		consumed, err := middleware.ConsumeLine(test.line)
+		assert.NoError(t, err, test.line)
+		assert.False(t, consumed, test.line)
+		assert.NotEqual(t, test.unexpectedState, middleware.state)
```
Comment (Waldz) [inline openvpn/middlewares/server/auth/middleware_test.go]: Should assert if `authenticator` was not called

File: openvpn/middlewares/server/auth/middleware_test.go [hunk 35]
```diff
@@ -0,0 +1,191 @@
...
+	var tests = []struct {
+		line          string
+		expectedState openvpn.State
+	}{
+		{">CLIENT:CONNECT,0,0", openvpn.STATE_AUTH},
+		{">CLIENT:ENV,password=12341234", openvpn.STATE_AUTH},
+		{">CLIENT:ENV,username=username1", openvpn.STATE_AUTH},
+		{">CLIENT:ENV,END", openvpn.STATE_UNDEFINED},
+	}
+	authFake := NewAuthenticatorFake()
```
Comment (Waldz) [inline openvpn/middlewares/server/auth/middleware_test.go]: Construct local callback function and You will be able to assert if `authFake` was called

File: openvpn/middlewares/server/auth/middleware_test.go [hunk 36]
```diff
@@ -0,0 +1,191 @@
...
+	connection := &fakeConnection{}
+	middleware.Start(connection)
+
+	for _, test := range tests {
+		consumed, err := middleware.ConsumeLine(test.line)
+		assert.NoError(t, err, test.line)
+		assert.True(t, consumed, test.line)
+		assert.Equal(t, test.expectedState, middleware.state)
+	}
+	authenticated, _ := authFake(middleware.lastUsername, middleware.lastPassword)
```
Comment (Waldz) [inline openvpn/middlewares/server/auth/middleware_test.go]: Dont call authentication but rather check that is was called by middleware

File: openvpn/middlewares/server/auth/middleware_test.go [hunk 37]
```diff
@@ -0,0 +1,191 @@
...
```
Comment (Waldz) [inline openvpn/middlewares/server/auth/middleware_test.go]: Same here

File: cmd/mysterium_server/command_run/factory.go [hunk 38]
```diff
@@ -66,8 +67,6 @@ func NewCommandWith(
 			return openvpn_session.NewManager(openvpn.NewClientConfig(
 				vpnServerIP,
 				filepath.Join(options.DirectoryConfig, "ca.crt"),
-				filepath.Join(options.DirectoryConfig, "client.crt"),
-				filepath.Join(options.DirectoryConfig, "client.key"),
```
Comment (donce) [inline cmd/mysterium_server/command_run/factory.go]: We should remove these files as well.

Comment (zolia) [reply cmd/mysterium_server/command_run/factory.go]: removed

**Pull request merged**